### PR TITLE
Split aq.function into aq.main and aq.subroutine

### PR DESCRIPTION
**Pull request by laurencap**
## Issue #, if available:
https://github.com/orgs/amazon-braket/projects/2/views/1?pane=issue&itemId=35858838

## Description of changes:
Split `aq.function` into `aq.main` and `aq.subroutine`.

I am planning to do a follow up PR to refactor api.py

### Example
input
```
@aq.subroutine
def bell(q0: int, q1: int):
    h(q0)
    cnot(q0, q1)

@aq.subroutine
def ghz_n(n: int):
    if n < 2:
        return
    for i in aq.range(n - 1):
        bell(i, i + 1)

@aq.main(num_qubits=6)
def main():
    ghz_n(6)
```
generated code
```
OPENQASM 3.0;
def bell(int[32] q0, int[32] q1) {
    h __qubits__[q0];
    cnot __qubits__[q0], __qubits__[q1];
}
def ghz_n(int[32] n) {
    if (n < 2) {
    } else {
        for int i in [0:n - 1 - 1] {
            bell(i, i + 1);
        }
    }
}
qubit[6] __qubits__;
ghz_n(6);
```

### Reasoning
**What's allowed at the global scope is not the same as what is allowed within a `subroutine`:** https://openqasm.com/language/scope.html#global-scope

Programming languages (including OpenQASM, Python, C++, Java) distinguish functions from the global scope in some way. For example, in Python:
```
print("Hello, World!")
```
is not the same as:
```
def wont_print():
    print("Hello, World!")
```

The same is true in OpenQASM:
```
h $0;
```
isn't the same as:
```
def subroutine():
    h $0;
```

It turns out, this distinction does matter to us even within AutoQASM. A great example of this is qubit declarations -- qubits are global resources and cannot be defined in subroutines. By separating out `main` from `subroutine`, we *completely eliminate* (!!) a failure mode for our customers. It's no longer possible to write AutoQASM code that declares 10 qubits in one function, but 20 qubits in a nested function call.

This separation also opens the door for us to do more careful error handling and program inspection wrt global scope

## Testing done:
Unit tests

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket/experimental/autoqasm/api.py [hunk 1]
```diff
@@ -227,25 +218,33 @@ def _convert_program(
...
         to the conversion process. Or, the oqpy variable returned from the converted function,
         if this is a subroutine conversion.
     """
-    # We will convert this function as a subroutine if we are already inside an
-    # existing conversion process (i.e., this is a subroutine call).
-    convert_as_subroutine = aq_program.in_active_program_conversion_context()
+    if is_subroutine and not aq_program.in_active_program_conversion_context():
+        raise errors.AutoQasmTypeError(
+            "Subroutines shouldn't be called directly. Please define an entry point "
+            "function, decorate it with '@autoqasm.main', and call your subroutine "
```
Comment (rmshaffer) [inline src/braket/experimental/autoqasm/api.py:232]: I think our docs are consistently using the `aq.` prefix
```suggestion
            "function, decorate it with '@aq.main', and call your subroutine "
```

File: src/braket/experimental/autoqasm/api.py [hunk 2]
```diff
@@ -227,25 +218,33 @@ def _convert_program(
...
+            "from within that function."
+        )
+    elif not is_subroutine and aq_program.in_active_program_conversion_context():
+        raise errors.AutoQasmTypeError(
+            f"Cannot call main function '{f.__name__}' from another main function. Did you mean "
+            "to use '@autoqasm.subroutine'?"
```
Comment (rmshaffer) [inline src/braket/experimental/autoqasm/api.py:238]:
```suggestion
            "to use '@aq.subroutine'?"
```

File: test/unit_tests/braket/experimental/autoqasm/test_api.py [hunk 3]
```diff
@@ -145,6 +145,10 @@ def test_sim_recursive_h_wrapper():
 
 
 def test_recursive_h():
+    @aq.main(num_qubits=6)
+    def main(n: int):
+        recursive_h(n)
```
Comment (rmshaffer) [inline test/unit_tests/braket/experimental/autoqasm/test_api.py:150]: Interesting, I guess we will not be able to support recursive `main` functions. Probably for the best, that doesn't make much sense anyway and it's very easy to wrap it as you have done here.

**Review by rmshaffer — APPROVED** [review]
This looks great! Even most of the tests were already written in a way that separating these was pretty straightforward - I feel that is further confirmation that this is a logical thing to do. Just a couple of trivial doc comments.

File: test/unit_tests/braket/experimental/autoqasm/test_api.py (hunk 3 above)
Comment (laurencap) [reply test/unit_tests/braket/experimental/autoqasm/test_api.py:150]: Agree. It feels like we might save our users from some weird bugs by disallowing this

**Pull request merged**